mailqueue.py

```diff
@@ -0,0 +1,181 @@
+        )
+
+        rows = self._db_cursor.fetchall()
+        if not rows:
+            return None
+
+        return [(r['rowid'], r['status']) for r in rows]
+
+    def mark_as_sent(self, envelope):
+        self._set_envelope_status(envelope, Status.SENT)
+
+    def mark_as_undeliverable(self, envelope):
+        self._set_envelope_status(envelope, Status.UNDELIVERABLE)
+
+    # TODO: split it into domain-specific methods, e.g. put_new_envelope() etc
+    def put(self, envelope):
+        self._execute_committing(
+            'INSERT INTO envelopes '
+            '(sender, recipients, destination_domain, message, next_attempt_at, submission_id, '
+            'status) '
+            'VALUES (?, ?, ?, ?, ?, ?, ?)',
+            (envelope.sender, ','.join(envelope.recipients), envelope.destination_domain,
+             str(envelope.message), self.clock.time(), envelope.submission_id, envelope.status)
+        )
+
+        return self._db_cursor.lastrowid
+
+    def schedule_retry_in(self, envelope, retry_after):
+        self._assert_envelope_has_id(envelope)
+        self._execute_committing(
+            "UPDATE envelopes SET next_attempt_at=?, delivery_attempts=? WHERE rowid=?",
+            (self.clock.time() + retry_after, envelope.delivery_attempts + 1, envelope.id)
+        )
+
+    @staticmethod
+    def _assert_envelope_has_id(envelope):
+        assert envelope.id, '{}: invalid envelope id "{}"'.format(envelope, envelope.id)
+
+    def _execute_committing(self, statement, *extra_args):
+        self._db_cursor.execute(statement, *extra_args)
+        self._db_conn.commit()
+
+    def _set_envelope_status(self, envelope, status):
+        self._assert_envelope_has_id(envelope)
+        self._execute_committing(
+            'UPDATE envelopes SET status="{}" WHERE rowid=?'.format(status), (envelope.id, ))
+
+
+# TODO: rename it to SynchronizedMailQueue
+class Manager:
+    """A proxy synchronizing access to MailQueue from multiple threads.
+
+    SQLite requires that the thread creating the database object is
+    the only one using it. This manager synchronizes access to methods
+    of MailQueue via blocking queues.
+    """
+
+    def __init__(self, **kwargs):
+        self._mail_queue_args = kwargs
+        self._manager_thread = threading.Thread(target=self._main_loop)
+        self._requests = queue.Queue(1)
+        self._responses = queue.Queue(1)
+
+        # XXX: factor out
+        methods = [name for name, _ in inspect.getmembers(MailQueue, inspect.isfunction)
+                   if not name.startswith('_')]
+
+        for method in methods:
+            assert method not in self.__class__.__dict__, \
+                "name clash: method '{}' already exists in {}".format(method,
+                                                                      self.__class__.__name__)
+            self.__dict__[method] = self.ProxiedMethod(method, self._requests, self._responses)
```

Comment:
Could you explain what it is for?

Reply:
Manager is a drop-in replacement for MailQueue. So we need it to have all the same methods as MailQueue.

One option is to manually implement each of them, but this means every change in MailQueue will lead to the same kind of change in Manager.

The other option (implemented here) is that Manager inspects the MailQueue class and dynamically creates all public methods (which proxy requests to MailQueue). This way Manager's methods are always up-to-date with MailQueue's and we don't have to manually replicate them each time there is a change in MailQueue.

Reply:
As far as I understood the Manager implements a synchronization mechanism for the MailQueue. I understand how it works but wouldn't things like `self._responses.put(getattr(mq, method)(*args))` be painful to debug when they throw an error?

Reply:
So, return values which indicate errors will be handled by callers of Manager.some_method(), so it shouldn't be an issue.

If a MailQueue method raises an exception, here's how it will look:

```
Traceback (most recent call last):
  File "/usr/lib/python3.6/threading.py", line 916, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.6/threading.py", line 864, in run
    self._target(*self._args, **self._kwargs)
  File "/app/mailqueue.py", line 174, in _main_loop
    self._responses.put(getattr(mq, method)(*args))
  File "/app/mailqueue.py", line 31, in raise_exception
    raise NotImplementedError
NotImplementedError
```
It does indicate MailQueue's method which raised the exception, so not all is lost (although we wouldn't know who issued the original call and with which arguments).

I don't see how we can improve diagnostics (without further complicating the mechanism, like passing a stack trace along with method arguments to the Manager thread) in a design where Manager runs in a separate thread.

SQLite has a requirement that a database connection should be only used by the thread which created it, thus warranting such design. There is a flag to relax this restriction, but I haven't looked closely into implications of its use.

The code for intra-thread communication with MailQueue is pretty simplistic and it provides *some* diagnostics which tend to be enough in practice (I usually write a failing end-to-end test before writing the code, so I have a chance to observe various ways how execution might fail as I gradually proceed with implementation).

So I suggest to leave it as is for now and get back to it later. Unless, of course, I'm missing something. What would be your suggestion?

Reply:
That was precisely my point. Although it looks succinct in the reality you'll have a hard time debugging it. Especially if that needs to be done quickly to mitigate production downtime.

Regarding the solution, if we decided to stick with SQLite (which is clearly limiting us) I would opt in for the following options:
* Use `check_same_thread=False` flag you mentioned. The downside is that you'll have to handle concurrency in the code.
* ~Implement task-agnostic client to the database that manages the serializations of the queries.~  It already exists: https://github.com/palantir/sqlite3worker
* I haven't looked deep enough but it seems the SQLite can be compiled with `SQLITE_THREADSAFE=1` option. Not sure if the python client supports it.

All these options have their own drawbacks and I agree that we can leave it as is. The important thing is that we strive for simplicity because it could be incredibly hard to debug another person's code under time pressure.

Reply:
I agree, if we stick with SQLite (which might be ok for the initial launch, but will probably have to be replaced with something else down the road), then `sqlite3worker` seems like the best way to go — good find!